x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/preprocessor/FullTextFunctionMapperPreprocessor.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.planner.mapper.preprocessor;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ResolvedIndices;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.Rewriteable;
+import org.elasticsearch.xpack.esql.expression.function.fulltext.FullTextFunction;
+import org.elasticsearch.xpack.esql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.planner.TranslatorHandler;
+import org.elasticsearch.xpack.esql.plugin.TransportActionServices;
+import org.elasticsearch.xpack.esql.session.IndexResolver;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Some {@link FullTextFunction} implementations such as {@link org.elasticsearch.xpack.esql.expression.function.fulltext.Match}
+ * will be translated to a {@link QueryBuilder} that require a rewrite phase on the coordinator.
+ * {@link FullTextFunctionMapperPreprocessor#preprocess(LogicalPlan, TransportActionServices, ActionListener)} will rewrite the plan by
+ * replacing {@link FullTextFunction} expression with new ones that hold rewritten {@link QueryBuilder}s.
+ */
+public class FullTextFunctionMapperPreprocessor implements MappingPreProcessor {
+
+    @Override
+    public void preprocess(LogicalPlan plan, TransportActionServices services, ActionListener<LogicalPlan> listener) {
+        Set<FullTextFunction> unresolved = fullTextFunctions(plan);
+        Set<String> indexNames = indexNames(plan);
+
+        if (indexNames == null || indexNames.isEmpty() || unresolved.isEmpty()) {
+            listener.onResponse(plan);
+            return;
+        }
+        QueryRewriteContext ctx = queryRewriteContext(services, indexNames);
+        FullTextFunctionsRewritable rewritable = new FullTextFunctionsRewritable(unresolved);
+        Rewriteable.rewriteAndFetch(
+            rewritable,
+            ctx,
+            listener.delegateFailureAndWrap((l, r) -> l.onResponse(updateQueryBuilders(plan, r.results())))
+        );
+    }
+
+    public LogicalPlan updateQueryBuilders(LogicalPlan plan, Map<FullTextFunction, QueryBuilder> newQueryBuilders) {
```

Comment:
Why isn't the rewrite applied on the spot and instead an associated map is used?

Reply:
I've reworked this bit. But my guess is that the intention was to avoid multiple tree passes.